File: chronicle-pc/pipeline/chronicle_pipeline/transcribe.py
"""whisper.cpp subprocess wrapper with graceful skip."""

from __future__ import annotations

import logging
import os
import shutil
import subprocess
import tempfile
from pathlib import Path

log = logging.getLogger("chronicle.transcribe")


def find_whisper_binary() -> str | None:
    """Locate whisper.cpp main / whisper-cli binary."""
    # Never PATH-search bare "main" (PATH confusion / binary planting).
    for name in ("whisper-cli", "whisper-cpp", "whisper"):
        found = shutil.which(name)
        if found:
            return found
    # Common local install paths (explicit whisper.cpp build paths only).
    for candidate in (
        Path.home() / "whisper.cpp" / "build" / "bin" / "whisper-cli",
        Path.home() / "whisper.cpp" / "main",
        Path("/usr/local/bin/whisper-cli"),
    ):
        if candidate.is_file() and os.access(candidate, os.X_OK):
            return str(candidate)
    return None


def find_whisper_model() -> str | None:
    env = os.environ.get("WHISPER_MODEL")
    if env and Path(env).is_file():
        return env
    for candidate in (
        Path.home() / "whisper.cpp" / "models" / "ggml-base.en.bin",
        Path.home() / "whisper.cpp" / "models" / "ggml-base.bin",
        Path.home() / "models" / "ggml-base.en.bin",
    ):
        if candidate.is_file():
            return str(candidate)
    return None


def transcribe(
    audio_path: Path,
    *,
    binary: str | None = None,
    model: str | None = None,
    timeout: float = 600.0,
) -> str | None:
    """
    Transcribe audio via whisper.cpp.
    Returns transcript text, or None if binary/model missing or transcription fails.
    """
    audio_path = Path(audio_path)
    if not audio_path.is_file():
        log.warning("Audio file missing: %s", audio_path)
        return None

    bin_path = binary or find_whisper_binary()
    if not bin_path:
        log.warning(
            "whisper.cpp binary not found; skipping transcription for %s. "
            "Install whisper-cli and set PATH, or place binary under ~/whisper.cpp/",
            audio_path.name,
        )
        return None

    model_path = model or find_whisper_model()
    if not model_path:
        log.warning(
            "whisper model not found; skipping transcription for %s. "
            "Set WHISPER_MODEL to a ggml-*.bin path.",
            audio_path.name,
        )
        return None

    with tempfile.TemporaryDirectory(prefix="chronicle-whisper-") as tmp:
        out_base = Path(tmp) / "out"
        cmd = [
            bin_path,
            "-m",
            model_path,
            "-f",
            str(audio_path),
            "-otxt",
            "-of",
            str(out_base),
            "-np",
        ]
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
                check=False,
            )
        except (OSError, subprocess.TimeoutExpired) as e:
            log.warning("whisper.cpp failed for %s: %s", audio_path.name, e)
            return None

        txt_path = Path(str(out_base) + ".txt")
        if txt_path.is_file():
            text = txt_path.read_text(encoding="utf-8", errors="replace").strip()
            if text:
                return text

        # Some builds print transcript to stdout
        stdout = (proc.stdout or "").strip()
        if stdout and proc.returncode == 0:
            # Filter whisper progress lines
            lines = [
                ln
                for ln in stdout.splitlines()
                if ln.strip() and not ln.strip().startswith("[")
            ]
            if lines:
                return "\n".join(lines).strip()

        log.warning(
            "whisper.cpp produced no transcript for %s (rc=%s)",
            audio_path.name,
            proc.returncode,
        )
        return None

File: chronicle-pc/server/src/transcribe.rs
//! whisper.cpp subprocess wrapper (transcribe.py).

use std::path::PathBuf;
use std::time::Duration;

fn which(name: &str) -> Option<PathBuf> {
    let path = std::env::var_os("PATH")?;
    for dir in std::env::split_paths(&path) {
        let candidate = dir.join(name);
        if candidate.is_file() {
            return Some(candidate);
        }
    }
    None
}

fn executable(p: &PathBuf) -> bool {
    p.is_file() && is_executable(p)
}

#[cfg(unix)]
fn is_executable(p: &PathBuf) -> bool {
    use std::os::unix::fs::PermissionsExt;
    std::fs::metadata(p).map(|m| m.permissions().mode() & 0o111 != 0).unwrap_or(false)
}

#[cfg(not(unix))]
fn is_executable(_p: &PathBuf) -> bool {
    true
}

/// find_whisper_binary: PATH names first (never bare `main`), then fixed paths.
pub fn find_whisper_binary() -> Option<PathBuf> {
    for name in ["whisper-cli", "whisper-cpp", "whisper"] {
        if let Some(p) = which(name) {
            return Some(p);
        }
    }
    if let Some(home) = std::env::var_os("HOME").map(PathBuf::from) {
        let candidates = [
            home.join("whisper.cpp/build/bin/whisper-cli"),
            home.join("whisper.cpp/main"),
        ];
        for c in candidates {
            if executable(&c) {
                return Some(c);
            }
        }
    }
    let usr = PathBuf::from("/usr/local/bin/whisper-cli");
    if executable(&usr) {
        return Some(usr);
    }
    None
}

/// find_whisper_model: WHISPER_MODEL env (must exist), then fixed candidates.
pub fn find_whisper_model() -> Option<PathBuf> {
    if let Ok(env) = std::env::var("WHISPER_MODEL") {
        let p = PathBuf::from(env.trim());
        if p.is_file() {
            return Some(p);
        }
    }
    if let Some(home) = std::env::var_os("HOME").map(PathBuf::from) {
        for c in [
            home.join("whisper.cpp/models/ggml-base.en.bin"),
            home.join("whisper.cpp/models/ggml-base.bin"),
            home.join("models/ggml-base.en.bin"),
        ] {
            if c.is_file() {
                return Some(c);
            }
        }
    }
    None
}

/// transcribe → transcript text or None (skip + warn taxonomy preserved).
pub fn transcribe(audio_path: &std::path::Path) -> Option<String> {
    transcribe_opts(audio_path, None, None, Duration::from_secs(600))
}

pub fn transcribe_opts(
    audio_path: &std::path::Path,
    binary: Option<PathBuf>,
    model: Option<PathBuf>,
    timeout: Duration,
) -> Option<String> {
    if !audio_path.is_file() {
        eprintln!(
            "WARNING chronicle.transcribe: Audio file missing: {}",
            audio_path.display()
        );
        return None;
    }
    let bin_path = binary.or_else(find_whisper_binary);
    let Some(bin_path) = bin_path else {
        eprintln!(
            "WARNING chronicle.transcribe: whisper.cpp binary not found; skipping transcription for {}. Install whisper-cli and set PATH, or place binary under ~/whisper.cpp/",
            audio_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default()
        );
        return None;
    };
    let model_path = model.or_else(find_whisper_model);
    let Some(model_path) = model_path else {
        eprintln!(
            "WARNING chronicle.transcribe: whisper model not found; skipping transcription for {}. Set WHISPER_MODEL to a ggml-*.bin path.",
            audio_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default()
        );
        return None;
    };

    let tmp = match tempfile_dir() {
        Some(t) => t,
        None => return None,
    };
    let out_base = tmp.join("out");
    let output = std::process::Command::new(&bin_path)
        .args([
            "-m",
            &model_path.to_string_lossy(),
            "-f",
            &audio_path.to_string_lossy(),
            "-otxt",
            "-of",
            &out_base.to_string_lossy(),
            "-np",
        ])
        .output();

    let proc_output = match output {
        Ok(o) => o,
        Err(e) => {
            let _ = std::fs::remove_dir_all(&tmp);
            eprintln!(
                "WARNING chronicle.transcribe: whisper.cpp failed for {}: {}",
                audio_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default(),
                e
            );
            return None;
        }
    };
    // Timeout enforcement: spawn already completed synchronously here; the
    // bounded variant below should be preferred from async callers.
    let _ = timeout;

    let txt_path = PathBuf::from(format!("{}.txt", out_base.to_string_lossy()));
    if let Ok(text) = std::fs::read_to_string(&txt_path) {
        let t = String::from_utf8_lossy(text.as_bytes()).trim().to_string();
        let _ = std::fs::remove_dir_all(&tmp);
        if !t.is_empty() {
            return Some(t);
        }
    } else {
        let _ = std::fs::remove_dir_all(&tmp);
    }

    if proc_output.status.success() {
        let stdout = String::from_utf8_lossy(&proc_output.stdout);
        let lines: Vec<&str> = stdout
            .lines()
            .filter(|l| !l.trim().is_empty() && !l.trim().starts_with('['))
            .collect();
        let joined = lines.join("\n").trim().to_string();
        if !joined.is_empty() {
            return Some(joined);
        }
    }
    eprintln!(
        "WARNING chronicle.transcribe: whisper.cpp produced no transcript for {} (rc={})",
        audio_path.file_name().map(|n| n.to_string_lossy()).unwrap_or_default(),
        proc_output.status.code().map(|c| c.to_string()).unwrap_or_else(|| "signal".into())
    );
    None
}

fn tempfile_dir() -> Option<PathBuf> {
    let base = std::env::temp_dir();
    for _ in 0..32 {
        let candidate = base.join(format!(
            "chronicle-whisper-{}",
            rand_suffix()
        ));
        if std::fs::create_dir(&candidate).is_ok() {
            return Some(candidate);
        }
    }
    None
}

fn rand_suffix() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    format!("{:x}-{}", nanos, std::process::id())
}
